server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MigrationUtil.java

```diff
@@ -97,99 +93,150 @@ public static synchronized void migrate(ProjectManager projectManager, CommandEx
+        }
+    }
+
+    private static void commitProjectMetadata(RepositoryUtil<ProjectMetadata> metadataRepo,
+                                              String projectName, ProjectMetadata metadata) {
+        try {
+            final Project project = metadataRepo.projectManager().get(projectName);
+            assert project != null;
+            if (!project.repos().exists(INTERNAL_REPOSITORY_NAME)) {
+                project.repos().create(INTERNAL_REPOSITORY_NAME);
+            }
+            metadataRepo.push(projectName, INTERNAL_REPOSITORY_NAME, author,
+                              "Add the metadata file",
+                              Change.ofJsonUpsert(METADATA_JSON, Jackson.valueToTree(metadata)))
+                        .toCompletableFuture().join();
+            logger.info("Project '{}' has been migrated", projectName);
+        } catch (Throwable cause) {
+            cause = Exceptions.peel(cause);
+            if (!(cause instanceof RedundantChangeException)) {
+                Exceptions.throwUnsafely(cause);
+            }
+        }
+    }
+
     private static Collection<Token> migrateTokens(ProjectManager projectManager,
                                                    CommandExecutor executor) {
+
         final RepositoryUtil<Tokens> tokensRepo =
                 new RepositoryUtil<>(projectManager, executor,
                                      entry -> convertWithJackson(entry, Tokens.class));
 
-        final Collection<LegacyToken> legacyTokens;
-
         final Project project = projectManager.get(INTERNAL_PROJECT_NAME);
-        if (project.repos().exists(LEGACY_TOKEN_REPO)) {
-            // Legacy tokens are stored in "dogma/tokens/token.json".
-            legacyTokens = project.repos().get(LEGACY_TOKEN_REPO)
-                                  .getOrNull(Revision.HEAD, LEGACY_TOKEN_JSON)
-                                  .thenApply(entry -> {
-                                      if (entry != null) {
-                                          return Jackson.<Map<String, LegacyToken>>convertValue(
-                                                  entry.content(), TOKEN_MAP_TYPE_REFERENCE).values();
-                                      } else {
-                                          return ImmutableList.<LegacyToken>of();
-                                      }
-                                  }).join();
-        } else {
-            legacyTokens = ImmutableList.of();
+
+        Tokens tokens = null;
+        // Check whether "dogma/main/tokens.json" exists or not in case of upgrading from 0.23.0.
+        // The difference between 0.23.0 and now is only the repository name.
+        if (project.repos().exists("main")) {
+            try {
+                tokens = tokensRepo.fetch(INTERNAL_PROJECT_NAME, "main", TOKEN_JSON)
+                                   .thenApply(HolderWithRevision::object)
+                                   .join();
+            } catch (Throwable ignore) {
```

Comment:
We should probably peel the exception and swallow only when it's `RepositoryNotFoundException` or `EntryNotFoundException`?

Reply:
We checked whether the repository exists before fetching. So it might not be raised. :-)